### PR TITLE
feat(inputs.mikrotik): Add plugin

### LINKED ISSUE #16080: Mikrotik input plugin
### Use Case

Gather metrics from Mikrotik RouterOS

### Expected behavior

The accumulator shall be populated with Mikrotik's metrics such as received/transfered bytes/packages etc

### Actual behavior

The accumulator shall be populated with Mikrotik's metrics such as received/transfered bytes/packages etc

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by pperminov**
## Summary

The idea is to be able to connect and gather metrics from the Mikrotik's RouterOS

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #16080

### CODE REVIEW

**Comment by DStrand1** [comment]
> This three issues I am having:
> 
> 1. https://github.com/influxdata/telegraf/actions/runs/11541885708/job/32124423506?pr=16081 I am highly confused. I ran super-linter locally and only issues I got are MD013 which are according to `.markdownlint.yml` shall be ignored.
> 2. https://app.circleci.com/pipelines/github/influxdata/telegraf/23202/workflows/acca2f7b-af70-4cc2-a827-736c98f95542/jobs/362160 This one looks like something not working properly. The script make_docs.sh when running locally formatting `telegraf/plugins/outputs/wavefront/README.md` so I believe it is the reason for the failure
> 3. https://app.circleci.com/pipelines/github/influxdata/telegraf/23202/workflows/acca2f7b-af70-4cc2-a827-736c98f95542/jobs/362162 not seems to be related at all

#16087 I believe should fix the first two, and the 3rd is an unrelated flaky test, so nothing needed here (other than a rebase/merge after the linked PR merges)

**Review by srebhan — CHANGES_REQUESTED** [review]
Thanks @s-r-engineer for your contribution! I want to encourage you to simplify your code! Why not defining known `tag` and `field` names of responses and decode the JSON responses to `map[string]string`? Then you can iterate over the map and decode the value accordingly? This would get rid of the reflection code and in general would simplify the code...

File: plugins/inputs/mikrotik/README.md
Comment (srebhan) [reply plugins/inputs/mikrotik/README.md:66]: It's hard to understand as you use `module` with different meanings I guess. How about
```suggestion
For each queried module there will be a different set of values and tags based on
the JSON structure that is returned from the REST endpoint. You can take
```

File: plugins/inputs/mikrotik/sample.conf
Comment (srebhan) [reply plugins/inputs/mikrotik/sample.conf:14]: No. :-) We use double-hashes for actual comments and single hashes for commented settings that do have a sensible default and/or are optional. Those commented _values_ must be of the form
```toml
  ## Allows to override the metric name
  # metric_name = "<default value for the metric name>"
```
i.e. they must show the default value of the setting.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [reply plugins/inputs/mikrotik/mikrotik.go:23]: Well it is also constant if you use the literal value in `AddMetric`! And it is mutable here and in `AddMetric` if someone edits the code, so what's the difference? Moreover, you now need to jump in the code if you are checking code around `AddMetric` and wonder what the metric name will be...

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [reply plugins/inputs/mikrotik/mikrotik.go:34]: Because that's how it is for all other plugins! We have to maintain 250+ plugins and searching for functions is pure fun, isn't it? ;-)

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [reply plugins/inputs/mikrotik/mikrotik.go:28]: 1. My review comments are not meant as insult or anything, I've asked that question because I wonder why this is necessary.
2. Then either make `shallWeParseTime` a member of the plugin and store the data there. The issue with this being global is that if you do have multiple instances of this plugin they will all share the same `durationParseTags` and therefore might break each other!

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [reply plugins/inputs/mikrotik/mikrotik.go:84]: I agree that they must know what they are doing, but it is a nice experience if you point a plugin to a device and it does _something_ e.g. collect the uptime or anything that is generally available or if it just collects _everything_.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [reply plugins/inputs/mikrotik/mikrotik.go:37]: Well if that is a fixed name for Mikrotik users, you shall leave the option as is. For me, as a non-mikrotik-dev, that's just an address when looking at the config...

File: plugins/inputs/mikrotik/parsers.go
Comment (srebhan) [reply plugins/inputs/mikrotik/parsers.go:73]: TBH, I would then prefer to wrap the parts checking for non-present parts like
```suggestion
	var duration time.Duration
	if matches[1] != "" {
		weeks, err := strToInt64(matches[1])
		if err != nil {
			return 0, err
		}
		duration += time.Duration(weeks*7*24)*time.Hour
	}
	if matches[2] != "" {
		days, err := strToInt64(matches[2])
		if err != nil {
			return 0, err
		}
		duration += time.Duration(days*24)*time.Hour
	}
	if matches[3] != "" {
		hours, err := strToInt64(matches[3])
		if err != nil {
			return 0, err
		}
		duration += time.Duration(hours)*time.Hour
	}
	if matches[4] != "" {
		minutes, err := strToInt64(matches[4])
		if err != nil {
			return 0, err
		}
		duration += time.Duration(minutes)*time.Minutes
	}
	if matches[5] != "" {
		seconds, err := strToInt64(matches[5])
		if err != nil {
			return 0, err
		}
		duration += time.Duration(seconds)*time.Seconds
	}
```

File: plugins/inputs/mikrotik/tools.go
Comment (srebhan) [reply plugins/inputs/mikrotik/tools.go:73]: Please view this from my reviewer point: If I see `strconv.ParseInt(str, 10, 64)` in the code, I _exactly_ know what it is doing. Now I see `strToInt64` in the code and need to navigate to the function to see what it does. Jumping around in the code hampers reading the code, so it has to be done with care...

File: plugins/inputs/mikrotik/tools.go
Comment (srebhan) [reply plugins/inputs/mikrotik/tools.go:69]: 1. This is not an abstraction, it is redirection.
2. Then please change the code _in the future_ and reintroduce this! 

We already do have enough code that complicates things just because it might have been helpful in the future where this future never happened! So please tailor your plugin to what it does __now__ and extend it in the future if you must!

**Comment by srebhan** [comment]
@s-r-engineer my review comments are that, comments about the code. They are not meant to criticize or insult you as a programmer or anyone else. If you understood it this way, please accept my apologies.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:28]: Please also make the `Username` a `config.Secret` so people can fill credentials from secret-stores if they need to.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:43]: Please just paste the code of `getSystemTags` in here. Indirections do make reviews and debugging harder especially in cases where you wrap functions with other functions for no benefit.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:48]: Please move this function before the `Start`.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:52]: Please just use `fmt.Errorf` here directly and use a more speaking error message. Please remember, these messages are user-facing! So maybe
```suggestion
		return errors.New("username must be set")
```

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:57]: Please move this to `init` to allow the user to actually _explicitly_ set the response timeout to zero aka infinite.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:72]: Please get rid of `wrapError`! If you return an error, the logger will add the plugin name anyway and if the error message is unique for `Init` there is no need for that function at all.

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:113]: Why not merge this into `Init`?

File: plugins/inputs/mikrotik/mikrotik.go
Comment (srebhan) [inline plugins/inputs/mikrotik/mikrotik.go:213]: Isn't this overly complex and corresponds to
```suggestion
	if resp.StatusCode != http.StatusOK {
		return data,  fmt.Errorf("server responded: %s (%d)", http.StatusText(resp.StatusCode)), resp.StatusCode)
	}
```

File: plugins/inputs/mikrotik/tools.go
Comment (srebhan) [inline plugins/inputs/mikrotik/tools.go:74]: Please use the code at the caller-side and get rid of these functions!

File: plugins/inputs/mikrotik/tools.go
Comment (srebhan) [reply plugins/inputs/mikrotik/tools.go:65]: But I do and our team has to maintain this code! So please use standard-stuff as much as possible as we cannot read into 300 different styles in our code base if we need to debug stuff!

File: plugins/inputs/mikrotik/types.go
Comment (srebhan) [inline plugins/inputs/mikrotik/types.go:6]: Please get rid of this indirection!

File: plugins/inputs/mikrotik/parsers.go
Comment (srebhan) [inline plugins/inputs/mikrotik/parsers.go:50]: You could create the metric here, can't you? This way you can get rid of the extra struct and/or directly add them to the accumulator.

File: plugins/inputs/mikrotik/parsers.go
Comment (srebhan) [inline plugins/inputs/mikrotik/parsers.go:16]: Why not filter the data in the loop below? This would save one iteration over all data.

File: plugins/inputs/mikrotik/parsers.go
Comment (srebhan) [inline plugins/inputs/mikrotik/parsers.go:12]: Why not keep the errors as errors and maybe directly log them _or_ if you need to join them in the end using `errors.Join`?

**Review by srebhan — COMMENTED** [review]
@s-r-engineer thanks for the update! We are coming closer... I still do have some comments especially around redirections in your code. Please do understand that we need to maintain this potentially "forever" and therefore need to enforce a certain style in the code otherwise debugging or extending this later will be much harder for us. We do have these examples in our codebase already and cannot extend that zoo infinitively.

**Comment by srebhan** [comment]
@s-r-engineer even though I do believe you that this is your intention, I've seen the opposite in the past, e.g. because people moved away from the hardware or due to other personal reasons.... Any way, I'm very happy with you maintaining that plugin!

Let me know as soon as you are ready for the next review round...

File: plugins/inputs/mikrotik/mikrotik.go
Comment (Hipska) [inline plugins/inputs/mikrotik/mikrotik.go:34]: Probably want to use client config from `plugins/common/http`?
```suggestion
	http.HTTPClientConfig
```

File: plugins/inputs/mikrotik/mikrotik.go
Comment (Hipska) [inline plugins/inputs/mikrotik/mikrotik.go:26]: Is it an idea to support an array of URLs?

File: plugins/inputs/mikrotik/mikrotik.go
Comment (Hipska) [inline plugins/inputs/mikrotik/mikrotik.go:55]:
```suggestion
		return errors.New("username is required")
```

File: plugins/inputs/mikrotik/mikrotik.go
Comment (Hipska) [inline plugins/inputs/mikrotik/mikrotik.go:71]:
```suggestion
			return fmt.Errorf("module %s does not exist or has a typo. Correct modules are: %s", selectedModule, getModuleNames())
```
These error prefixes are not needed, just make sure the rest of the message is specific enough.

**Review by Hipska — CHANGES_REQUESTED** [review]

**Comment by srebhan** [comment]
@s-r-engineer would you be OK if I work on the code to get it to a mergable state? You are so close and there are only small things left to do... I would do it on your branch if you provide write access or in a separate branch if you are ok with it...

**Comment by srebhan** [comment]
We will respect your choice of course and not work on the plugin. If you decide you want to add the plugin upstream one day, please do not hesitate to reopen the PR or add a new one.

I'm sad we seem to have scared you away from adding this plugin and I would be interested in learning what we did wrong...

**Pull request closed without merging**
